src/Symfony/Contracts/HttpClient/HttpClientInterface.php

```diff
@@ -0,0 +1,82 @@
+        'body' => '',           // array|string|resource|\Traversable|\Closure - the callback SHOULD yield a string
+                                //   smaller than the amount requested as argument; the empty string signals EOF; when
+                                //   an array is passed, it is meant as a form payload of field names and values
+        'json' => null,         // array|\JsonSerializable - when set, implementations MUST set the "body" option to
+                                //   the JSON-encoded value and set the "content-type" headers to a JSON-compatible
+                                //   value it is they are not defined - typically "application/json"
+        'max_redirects' => 20,  // int - the maximum number of redirects to follow - any "Authorization" headers MUST
+                                //       NOT follow except for the initial host name as given in the original URL
+        'http_version' => null, // string - defaults to the best supported version, typically 1.1 or 2.0
+        'base_uri' => null,     // string - the URI to resolve relative URLs, following rules in RFC 3986, section 2
+        'buffer' => true,       // bool - whether the content of the response should be buffered or not
+        'on_progress' => null,  // callable - called on download/upload at least once per second with arguments
+                                //            int $dlNow, int $dlSize and array $info; any exceptions thrown from the
+                                //            callback MUST abort the request
+        'resolve' => [],        // string[] - a map of host to IP address that SHOULD replace DNS resolution
+        'proxy' => null,        // string - by default, the proxy-related env vars handled by curl SHOULD be honored
+        'timeout' => null,      // float - the inactivity timeout - defaults to ini_get('default_socket_timeout')
+        'bindto' => '0',        // string - the interface or the local socket to bind to
+        'verify_peer' => true,  // see https://php.net/context.ssl for the following options
+        'verify_host' => true,
+        'cafile' => null,
+        'capath' => null,
+        'local_cert' => null,
+        'local_pk' => null,
+        'passphrase' => null,
+        'ciphers' => null,
+        'peer_fingerprint' => null,
+        'capture_peer_cert_chain' => false,
+    ];
+
+    /**
+     * Requests an HTTP resource.
+     *
+     * Responses SHOULD be lazy, but their status code MUST be
+     * checked even if none of their public methods are called.
+     *
+     * @param string   $url     The URL of the resource
+     * @param iterable $options The options of the request
+     *
+     * @throws TransportExceptionInterface When an unsupported option is passed or
+     *                                     if any error happens at the transport layer
+     */
+    public function request(string $method, string $url, iterable $options = []): ResponseInterface;
```

Comment:
in my opinion, the interface is a very leaky abstraction. rather than having all these things configured on a per-request basis, they should be configured on the client instance. having the options here tells people to leak all kind of details into their code interacting with the http client, instead of having that managed via DI on the http client itself.

if, say, there is one host where we do not want ssl verification because of an outdated certificate, i'd rather have a meta client that choses the correctly configured client based on a request detail like the hostname than littering this decision over all my code that interacts with the http client.

this is the reasoning we took for the PSR-18 request method, and if we would do the same here it would be possible to write an adapter from the symfony http client to use any psr-18 client. (the adapter in this PR is for using the symfony client as PSR-18 client, which is the other direction)

Reply:
well, headers and body also appear as options in that signature for instance

Reply:
yes, i am not arguing against headers and body, but against all the nitty-gritty http protocol options.

Reply:
hi @dbu, I have a different opinion here: all concrete HTTP clients have to deal with these settings. The defined options *do* abstract these. In PSR-18, you decided to not expose anything about the transport layer. This effectively forbids from controlling these settings in any interoperable ways. Your SSL example is a really good one: there is no way with PSR-18 itself to implement the logic you describe. I made a different choice and decided to expose the layer via the abstraction provided by the options. This will allow implementing interoperable decorators that do e.g. conditional SSL routing, and much more. IMHO, users need power tools, we should not restrict what they can do. If they use an HTTP client, it's legit that they can decide to control every aspect of it. Trying to save them from themselves could lead to the opposite effect: they will hack around the limitations and bound themselves to one implementation. That's exactly what happened to the Blackfire Player: its implementation is now bound to Guzzle because it needs DSN pre-resolution for security reasons; escaping the interop promise.

Reply:
i agree that being overly limiting is a problem. however, this is not good flexibility but prompting bad patterns. having an $options parameter means that the configuration of how you want your client to work is carried throughout your application or hardcoded in all places that send requests. 
using correctly configured clients is a much better way to be sure all requests the application does are done with the correct settings. my example with ssl was to illustrate that even if you have a complicated situation, you can still inject one single client to your code and handle things in the client. the common use case however, will be that you send requests to one single server and want the same settings for all requests.

if i understand correctly, the list of options is not closed to what is mentioned in the interface, but could also be additional things. clients will not implement all of them, and some of the options are likely not well defined. if everything in the end uses lib_curl, you will have that as a de-facto standard, but clients that do not build on curl will likely not respect that.

Reply:
both ways are and should be supported: default options allow injecting preconfigured clients - of course, that's out of scope as no contracts can cover DI configuration.
Enabling decorators to precisely control HTTP-transport settings in an interoperable way is a goal of the proposed contracts. The list of options is in the contracts precisely because I want to make it a requirement to support them all - *or throw*. **All** options are supported by both native and curl implementations. There is zero binding to curl here. Note that *not* supporting an option is also possible, but then a client MUST throw. That applies to custom options of course so that if you rely on one of them, you'll bind yourself to one specific implementation, but at least you'll know when swapping the implementation breaks your need (very alike using methods that are only on implementations and not on some interfaces.)

Reply:
> The list of options is in the contracts precisely because I want to make it a requirement to support them all - or throw.

i either missed that part or its not yet explicitly written out. when going with this approach, i think its a very good point, because it makes things a lot more reliable. 
do i get it correctly that the list is open, so additional options might be supported, but clients MUST throw an exception if they don't know any option that they have been passed?

i am still not convinced that this is the right approach. but with that clarification, the whole thing looks a lot better to me :+1:

Reply:
My communication style is never explicit enough, thanks for helping me brain-dump :) I'm sure there is room for clarifications by PRs after merging. You're right about added and unsupported options, that's what I meant by the description next to `@throws`.